Core/HLE/sceNetAdhoc.cpp

```diff
@@ -113,8 +120,457 @@ void __NetAdhocShutdown() {
 	}
 }
 
+static void __AdhocctlNotify(u64 userdata, int cyclesLate) {
+	SceUID threadID = userdata >> 32;
+	int uid = (int)(userdata & 0xFFFFFFFF);
+
+	s64 result = 0;
+	u32 error = 0;
+
+	SceUID waitID = __KernelGetWaitID(threadID, WAITTYPE_NET, error);
+	if (waitID == 0 || error != 0)
+		return; // FIXME: Is it safe to exit here like this without re-scheduling the event? Will this event be triggered again? What will happen to the result i might want to change if exited here?
```

Comment:
Well, it won't automatically.  But if the thread is not waiting no that, something has gone wrong and you probably don't want it to.

Most common cause would be the game calling `sceKernelReleaseWaitThread()`.

-[Unknown]

Reply:
During my test it was woken-up by hleDelayResult on the AdhocMatching loop hle, so i'm replacing it with sceKernelDelayThread and didn't get woken-up randomly anymore.

What i want to know it, in the case the waiting sceNet function got woken up by something else, what kind of return value will it get? was it the normal return value from that function will be used?

For example:
```
int sceNetSomething(){
...
	CoreTiming::ScheduleEvent(usToCycles(100), adhocSocketNotifyEvent, param);
	__KernelWaitCurThread(WAITTYPE_NET, socketId, 0, 0, false, reason);

	// Fallback return value?
	return ERROR_NET_ADHOC_TIMEOUT;
}
```
i assume the ERROR_NET_ADHOC_TIMEOUT will be used as return value in the case the thread got woken up unexpectedly, is that a correct assumption?

Reply:
The "normal" return value is discarded.  Nothing will use that.

`SCE_KERNEL_ERROR_RELEASE_WAIT` is the return value if `sceKernelReleaseWaitThread` is used.  If you use `hleDelayResult` on a thread that was already waiting on something else, its new result will be whatever you specified in the `hleDelayResult()` call.

`__KernelGetWaitID()` actually returns -1 on error, so technically 0 just means you specified 0 as the wait id in `__KernelWaitCurThread()` or etc.

PSP threads won't be woken unexpectedly.  There's no such thing as spurious wakeup in our HLE thread management.

-[Unknown]